### PR TITLE
XRC-01 Contract Standard

### LINKED ISSUE #421: XRC-01 Contract Standard
**Brief of your idea**
XRCs  are technical documents used by smart contract developers at XuperChain.

XRC-01 is a multi token standard used for smart contracts on the XuperChain blockchain. It's can include combination of fungible and non-fungible tokens in one contract. The default asset transferor is the contract initiator.

Smart contracts implementing the XRC-01 standard can just implement the interface they needed. The main interfaces of this standard are as following:

- issue
- authorization
- withdraw_authorization 
- transfer
- transfer_from
- get_balance
- get_authorized
- owner_of
- authorize_infos
- authorized_infos

**What could be better if XuperChain using your idea**
Users can map offline assets to online assets easily. 

**Are there any side effects?**
Use table to store asset information. May cause write amplification.

### PR DESCRIPTION
**Pull request by qizheng09**
## Description

XRCs are technical documents used by smart contract developers at XuperChain.

XRC-01 is a multi token standard used for smart contracts on the XuperChain blockchain. It's can include combination of fungible and non-fungible tokens in one contract. The default asset transferor is the contract initiator.

Fixes #421 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Brief of your solution

Smart contracts implementing the XRC-01 standard can just implement the interface they needed.
**`XRC/xrc01/src/main/xrc01_e1.cc` is the example.**

 The main interfaces of this standard are as following:

- issue
- authorization
- withdraw_authorization 
- transfer
- authorization_transfer
- get_balance
- get_authorized
- owner_of
- authorize_infos
- authorized_infos

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

### CODE REVIEW

File: contractsdk/cpp/table/table.tpl.h [hunk 1]
```diff
@@ -189,6 +190,43 @@ bool Table<T>::put(T t) {
     return true;
 }
 
+template <typename T>
+bool Table<T>::update(T t) {
+    const std::string& key = t.rowkey();
+    const std::string& all_key = make_key(PREFIX_ROWKEY,key);
+    std::string value;
+    if (!_ctx->get_object(all_key, &value)) {
+        std::cout << "value not found: " << std::endl;
```
Comment (duanbing) [inline contractsdk/cpp/table/table.tpl.h]: clear the debug print here.

File: contractsdk/cpp/XRC/xrc01/src/main/xrc01.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include "xchain/xchain.h"
+#include "table/types.h"
+#include "table/table.tpl.h"
+#include "pb/xrc01.pb.h"
+
+// XRC01 is the implement of XRC-01 multi token standard
+class XRC01 {
```
Comment (duanbing) [inline contractsdk/cpp/XRC/xrc01/src/main/xrc01.h]: if this is our proposal for token issuing, we should provide an interface first, then is our implementation.

File: contractsdk/cpp/XRC/xrc01/src/main/xrc01_e1.cc [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    }    
+    ctx->ok("withdraw_authorization succeed");  
+}
+
+DEFINE_METHOD(XRC01_E1, transfer) {
+    xchain::Context* ctx = self.context();
+    XRC01 xrc01(ctx);
+
+    const std::string& to = ctx->arg("to");
+    uint64_t token_id = std::stoll(ctx->arg("token_id"));
```
Comment (duanbing) [inline contractsdk/cpp/XRC/xrc01/src/main/xrc01_e1.cc]: check overflow. 
"strtoll" is safer, but not enough though.  refer to the wiki I shared in the hi group.

File: contractsdk/cpp/XRC/xrc01/src/main/xrc01.h (hunk 2 above)
Comment (qizheng09) [reply contractsdk/cpp/XRC/xrc01/src/main/xrc01.h]: Good idea, but this has been reviewed internally for several times.  In the future, we can define a process to propose XRC standard. I will write a wiki for this.

File: contractsdk/cpp/XRC/xrc01/src/main/xrc01_e1.cc (hunk 3 above)
Comment (qizheng09) [reply contractsdk/cpp/XRC/xrc01/src/main/xrc01_e1.cc]: ok, I will fix it by providing a safe `stoull` in `safemath` package.

File: contractsdk/cpp/table/table.tpl.h (hunk 1 above)
Comment (qizheng09) [reply contractsdk/cpp/table/table.tpl.h]: done

File: contractsdk/cpp/XRC/xrc01/src/main/xrc01.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    public: 
+        XRC01(xchain::Context* ctx): _token(ctx, "token"), _asset_info(ctx, "asset_info"), _authorize_info(ctx, "_authorize_info") {
```
Comment (yucaowang) [inline contractsdk/cpp/XRC/xrc01/src/main/xrc01.h]: Consider following C++ code style to keep each line no longer than 80 chars.

Comment (qizheng09) [reply contractsdk/cpp/XRC/xrc01/src/main/xrc01.h]: fix it

**Pull request merged**